JoinColony/colonyDapp
# Refactor contract parser validation

## Description

This PR refactors the contract parser validation to accomodate arrays, and all of the various uint/int and byte sizes. It also moves the validation schema logic into its own file as it is getting rather large.

Please see the end of the issue for some relevant observations about performance.

**New stuff** ✨

* Validation helpers
* Add `formatjs/intl` package to format messages outside of React components (facilitates extraction of validation schema)

**Changes** 🏗

* Extracted validation schema into its own file 
* Refactor contract function parser validation

To test:
1. Go to Contract Interaction in Safe Control
2. Use the abi below if you want or use your own
3. Test out the validation. Should permit arrays and have the correct error messages. 

BytesX needs to be length x * 2 + 2 and start with 0x
intX needs to be within  +-((2 ** (X - 1)) - 1)
uintX needs to be 0 or above and less than 2 ** X  - 1
Arrays need to start with '[', end with ']', and each item needs to validate correctly
bool needs to be 'true' or 'false'
address needs to begin '0x' and be a valid address

 
## Test ABI

The following ABI covers a good cross-section of the added validation.

```
[
	{
		"inputs": [
			{
				"internalType": "bytes4[]",
				"name": "a",
				"type": "bytes4[]"
			},
			{
				"internalType": "bytes30[]",
				"name": "b",
				"type": "bytes30[]"
			},
			{
				"internalType": "bytes16",
				"name": "c",
				"type": "bytes16"
			},
			{
				"internalType": "address[]",
				"name": "d",
				"type": "address[]"
			},
			{
				"internalType": "address",
				"name": "e",
				"type": "address"
			},
			{
				"internalType": "uint56",
				"name": "f",
				"type": "uint56"
			}
		],
		"name": "one",
		"outputs": [],
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"inputs": [
			{
				"internalType": "uint200[]",
				"name": "g",
				"type": "uint200[]"
			},
			{
				"internalType": "uint256[]",
				"name": "h",
				"type": "uint256[]"
			},
			{
				"internalType": "int256[]",
				"name": "i",
				"type": "int256[]"
			},
			{
				"internalType": "int8",
				"name": "j",
				"type": "int8"
			},
			{
				"internalType": "bool[]",
				"name": "k",
				"type": "bool[]"
			},
			{
				"internalType": "bool",
				"name": "l",
				"type": "bool"
			},
			{
				"internalType": "string[]",
				"name": "m",
				"type": "string[]"
			},
			{
				"internalType": "string",
				"name": "n",
				"type": "string"
			}
		],
		"name": "two",
		"outputs": [],
		"stateMutability": "nonpayable",
		"type": "function"
	}
]
```

## Discussion

The `Safe Control` form can be of arbitrary size given the user's ability to add transactions. Currently, the entire form validates when any of the inputs change. Validation takes time to run, and with the increasing complexity added here, I'm starting to notice a bit of input lag. For example, hold down a single key in an input field and it gets a bit choppy. 

Performance will degrade with form size, since a larger form requires more validation to be run, and could degrade noticeably if, say, we had a large form full of contract interaction transactions, all of which had many inputs with many large arrays. 

I tested this out as a proof of concept. In the gif below, I have 5 Contract Interaction Tabs open, all with the abi in this issue, but only the first one contains any user input. You can see that there's a lag as I hold down the "1" key, and that's only doing so for a second or so (the longer you hold it down for, the worse it gets).
 
![111](https://user-images.githubusercontent.com/64402732/192004652-c876f15b-2ef2-48e4-9ea8-02d13f705dbe.gif)

I think this indicates a problem with how we're validating this form. I haven't thought too much about the solution since I wanted to first confirm my observations. However, my first thought is that we validate inputs at the field level onchange (or even onblur), and only run entire form validation once when the form is submitted . The easiest way to do this may to be alter the `Input` component to accept a `validateOnChange/OnBlur` prop and call `validateField()` inside the input's `onChange/Blur` function if it is true.

Thoughts very welcome.

Resolves #3827
